File: app/lib/core/e2ee/device_profile.dart
import 'dart:convert';
import 'dart:typed_data';

const _profileVersion = 1;

class E2eeDeviceProfile {
  const E2eeDeviceProfile({
    required this.deviceId,
    required this.deviceLabel,
    required this.registered,
    required this.keyPackagePublished,
    this.lastControlSequences = const {},
  });

  final String deviceId;
  final String deviceLabel;
  final bool registered;
  final bool keyPackagePublished;
  final Map<String, int> lastControlSequences;

  E2eeDeviceProfile copyWith({
    bool? registered,
    bool? keyPackagePublished,
    Map<String, int>? lastControlSequences,
  }) => E2eeDeviceProfile(
    deviceId: deviceId,
    deviceLabel: deviceLabel,
    registered: registered ?? this.registered,
    keyPackagePublished: keyPackagePublished ?? this.keyPackagePublished,
    lastControlSequences: lastControlSequences ?? this.lastControlSequences,
  );

  Uint8List encode() => Uint8List.fromList(
    utf8.encode(
      jsonEncode({
        'version': _profileVersion,
        'device_id': deviceId,
        'device_label': deviceLabel,
        'registered': registered,
        'key_package_published': keyPackagePublished,
        'last_control_sequences': lastControlSequences,
      }),
    ),
  );

  static E2eeDeviceProfile decode(List<int> value) {
    final data = jsonDecode(utf8.decode(value));
    if (data is! Map<String, dynamic> ||
        data['version'] != _profileVersion ||
        data['device_id'] is! String ||
        (data['device_id'] as String).trim().isEmpty ||
        data['device_label'] is! String ||
        data['registered'] is! bool ||
        data['key_package_published'] is! bool ||
        data['last_control_sequences'] is! Map) {
      throw const FormatException('E2EE 设备档案格式无效');
    }
    final sequences = (data['last_control_sequences'] as Map).map((key, value) {
      if (key is! String || value is! int || value < 0) {
        throw const FormatException('E2EE 控制消息游标无效');
      }
      return MapEntry(key, value);
    });
    return E2eeDeviceProfile(
      deviceId: data['device_id'] as String,
      deviceLabel: data['device_label'] as String,
      registered: data['registered'] as bool,
      keyPackagePublished: data['key_package_published'] as bool,
      lastControlSequences: Map.unmodifiable(sequences),
    );
  }
}

File: h5-app/src/e2ee/device-profile.ts
const PROFILE_VERSION = 1;

export interface E2eeDeviceProfile {
  deviceId: string;
  deviceLabel: string;
  registered: boolean;
  keyPackagePublished: boolean;
  deviceStatus?: 'active' | 'pending_approval';
  lastControlSequences: Record<string, number>;
  lastCommitMessageIds: Record<string, string>;
}

export const encodeDeviceProfile = (profile: E2eeDeviceProfile) => new TextEncoder().encode(JSON.stringify({
  version: PROFILE_VERSION,
  device_id: profile.deviceId,
  device_label: profile.deviceLabel,
  registered: profile.registered,
  key_package_published: profile.keyPackagePublished,
  device_status: profile.deviceStatus ?? 'active',
  last_control_sequences: profile.lastControlSequences,
  last_commit_message_ids: profile.lastCommitMessageIds,
}));

export const decodeDeviceProfile = (value: Uint8Array): E2eeDeviceProfile => {
  const data = JSON.parse(new TextDecoder('utf-8', { fatal: true }).decode(value)) as Record<string, unknown>;
  if (data.version !== PROFILE_VERSION
    || typeof data.device_id !== 'string'
    || !data.device_id.trim()
    || typeof data.device_label !== 'string'
    || typeof data.registered !== 'boolean'
    || typeof data.key_package_published !== 'boolean'
    || (data.device_status != null && data.device_status !== 'active' && data.device_status !== 'pending_approval')
    || !isRecord(data.last_control_sequences)
    || (data.last_commit_message_ids != null && !isRecord(data.last_commit_message_ids))) {
    throw new Error('E2EE 设备档案格式无效');
  }
  const sequences = Object.fromEntries(Object.entries(data.last_control_sequences).map(([key, item]) => {
    if (!Number.isSafeInteger(item) || Number(item) < 0) throw new Error('E2EE 控制消息游标无效');
    return [key, Number(item)];
  }));
  const commitIds = Object.fromEntries(Object.entries(
    isRecord(data.last_commit_message_ids) ? data.last_commit_message_ids : {},
  ).map(([key, item]) => {
    if (!key.trim() || typeof item !== 'string' || !item.trim()) throw new Error('E2EE Commit 索引无效');
    return [key, item];
  }));
  return {
    deviceId: data.device_id,
    deviceLabel: data.device_label,
    registered: data.registered,
    keyPackagePublished: data.key_package_published,
    deviceStatus: data.device_status === 'pending_approval' ? 'pending_approval' : 'active',
    lastControlSequences: sequences,
    lastCommitMessageIds: commitIds,
  };
};

const isRecord = (value: unknown): value is Record<string, unknown> => (
  Boolean(value && typeof value === 'object' && !Array.isArray(value))
);
